File: packages/node/src/query/present.ts
/** Plain-text presenters for the CLI query mirrors (search/read/neighbors/overview).
 *
 * The mirrors are thin: they call the very same payload builders the MCP tools and
 * REST use, then render for a human here. `--json` skips these entirely and prints
 * the raw payload — the same shape a machine gets over MCP. Ports query/present.py.
 */

export function toJson(payload: unknown): string {
  return JSON.stringify(payload, null, 2);
}

interface Counts {
  docs?: number;
  edges?: number;
  tags?: number;
  orphans?: number;
  ghosts?: number;
}

function countsLine(counts: Counts): string {
  return (
    `${counts.docs ?? 0} docs · ${counts.edges ?? 0} links · ` +
    `${counts.tags ?? 0} tags · ${counts.orphans ?? 0} orphans · ${counts.ghosts ?? 0} ghosts`
  );
}

interface OverviewDoc {
  path: string;
  title: string;
  description: string | null;
}

export function presentOverview(p: Record<string, unknown>): string {
  const counts = (p["counts"] ?? {}) as Counts;
  const tiers = (p["tiers"] ?? {}) as Record<string, unknown>;
  const lines = [
    `brain: ${String(p["bundle"] ?? "")}`,
    `counts: ${countsLine(counts)}`,
    "tiers: " +
      Object.entries(tiers)
        .map(([k, v]) => `${k} ${String(v)}`)
        .join(" · "),
  ];
  for (const group of (p["tree"] ?? []) as Array<{ group: string; docs: OverviewDoc[] }>) {
    lines.push("");
    lines.push(`${group.group}/`);
    for (const doc of group.docs) {
      const desc = doc.description ? ` — ${doc.description}` : "";
      lines.push(`  ${doc.path}  ${doc.title}${desc}`);
    }
  }
  if (p["truncated"]) {
    lines.push("");
    lines.push("(listing trimmed — pass --json or a bigger brain for the full tree)");
  }
  return lines.join("\n");
}

interface Hit {
  path: string;
  title: string;
  description: string | null;
  why?: string;
  snippet?: string | null;
}

export function presentSearch(p: Record<string, unknown>, query: string): string {
  const hits = (p["hits"] ?? []) as Hit[];
  let header = `${hits.length} hits for '${query}'`;
  const used = ((p["used_modes"] ?? []) as string[]).join(", ");
  if (used) header += ` (mode: ${used})`;
  if (p["degraded_from"]) header += ` [degraded from ${String(p["degraded_from"])}]`;
  const lines = [header];
  for (const hit of hits) {
    const desc = hit.description ? ` — ${hit.description}` : "";
    const why = hit.why ? `  (${hit.why})` : "";
    lines.push(`  ${hit.path}  ${hit.title}${desc}${why}`);
    if (hit.snippet) lines.push(`      ↳ ${hit.snippet}`);
  }
  if (hits.length === 0) lines.push("  (no hits — try `overview` for the whole brain)");
  return lines.join("\n");
}

function presentResolutionMiss(p: Record<string, unknown>): string | null {
  if ("disambiguation" in p) {
    const lines = ["several docs match — name one exactly:"];
    for (const d of p["disambiguation"] as Array<{ path: string; title: string }>) {
      lines.push(`  ${d.path}  ${d.title}`);
    }
    return lines.join("\n");
  }
  if ("error" in p) {
    const lines = [String(p["error"])];
    const suggestions = (p["suggestions"] ?? []) as string[];
    if (suggestions.length > 0) {
      lines.push("did you mean:");
      for (const s of suggestions) lines.push(`  ${s}`);
    }
    return lines.join("\n");
  }
  return null;
}

export function presentRead(p: Record<string, unknown>): string {
  const miss = presentResolutionMiss(p);
  if (miss !== null) return miss;
  const frontmatter = (p["frontmatter"] ?? {}) as Record<string, unknown>;
  const title = frontmatter["title"];
  const heading = title ? `${String(p["path"])} — ${String(title)}` : String(p["path"]);
  const lines = [heading, "", String(p["content"] ?? "").replace(/\n+$/, "")];
  const neighbors = (p["neighbors"] ?? {}) as { in?: Array<{ path: string }>; out?: Array<{ path: string }> };
  const incoming = (neighbors.in ?? []).map((n) => n.path).join(", ");
  const outgoing = (neighbors.out ?? []).map((n) => n.path).join(", ");
  if (incoming || outgoing) {
    lines.push("");
    lines.push(`neighbors: in [${incoming}] out [${outgoing}]`);
  }
  if (p["truncated"]) {
    lines.push("");
    lines.push("(content trimmed — pass --json or read specific sections for the rest)");
  }
  return lines.join("\n");
}

/** The brainpick show result (spec/95): what the live server broadcast. */
export function presentShow(p: Record<string, unknown>): string {
  if ("error" in p) return String(p["error"]);
  const shown = (p["shown"] ?? 0) as number;
  const lines = [`presented (seq ${String(p["seq"])}): ${shown} node(s) spotlighted in every open UI`];
  const dropped = (p["dropped"] ?? []) as string[];
  if (dropped.length > 0) lines.push(`dropped (unresolved): ${dropped.join(", ")}`);
  return lines.join("\n");
}

interface NeighborNode {
  path: string;
  title: string;
  description: string | null;
  distance: number;
}

export function presentNeighbors(p: Record<string, unknown>): string {
  const miss = presentResolutionMiss(p);
  if (miss !== null) return miss;
  const center = String(p["center"]);
  let header = `neighbors of ${center}`;
  if (p["degraded_from"]) header += ` [degraded from ${String(p["degraded_from"])}]`;
  const lines = [header];
  const others = ((p["nodes"] ?? []) as NeighborNode[]).filter((n) => n.path !== center);
  for (const node of others) {
    const desc = node.description ? ` — ${node.description}` : "";
    lines.push(`  d${node.distance}  ${node.path}  ${node.title}${desc}`);
  }
  if (others.length === 0) lines.push("  (no neighbors — an orphan, or nothing links here yet)");
  const edges = (p["edges"] ?? []) as Array<{ source: string; target: string; kind: string }>;
  if (edges.length > 0) {
    lines.push("edges:");
    for (const e of edges) lines.push(`  ${e.source} → ${e.target} (${e.kind})`);
  }
  return lines.join("\n");
}

File: packages/python/src/brainpick/query/present.py
"""Plain-text presenters for the CLI query mirrors (search/read/neighbors/overview).

The mirrors are thin: they call the very same payload builders the MCP tools and
REST use, then render for a human here. `--json` skips these entirely and prints
the raw payload — the same shape a machine gets over MCP.
"""
from __future__ import annotations

import json


def to_json(payload: dict) -> str:
    """The machine face: the payload verbatim, pretty-printed and UTF-8."""
    return json.dumps(payload, ensure_ascii=False, indent=2)


def _counts_line(counts: dict) -> str:
    return (
        f"{counts.get('docs', 0)} docs · {counts.get('edges', 0)} links · "
        f"{counts.get('tags', 0)} tags · {counts.get('orphans', 0)} orphans · "
        f"{counts.get('ghosts', 0)} ghosts"
    )


def present_overview(payload: dict) -> str:
    lines = [
        f"brain: {payload.get('bundle', '')}",
        f"counts: {_counts_line(payload.get('counts', {}))}",
        "tiers: " + " · ".join(f"{k} {v}" for k, v in payload.get("tiers", {}).items()),
    ]
    if "similarity_gaps_open_count" in payload:
        lines.append(f"similarity gaps: {payload['similarity_gaps_open_count']} open")
    for group in payload.get("tree", []):
        lines.append("")
        lines.append(f"{group['group']}/")
        for doc in group["docs"]:
            desc = f" — {doc['description']}" if doc.get("description") else ""
            lines.append(f"  {doc['path']}  {doc['title']}{desc}")
    if payload.get("truncated"):
        lines.append("")
        lines.append("(listing trimmed — pass --json or a bigger brain for the full tree)")
    return "\n".join(lines)


def present_search(payload: dict, query: str) -> str:
    hits = payload.get("hits", [])
    header = f"{len(hits)} hits for '{query}'"
    used = ", ".join(payload.get("used_modes", []))
    if used:
        header += f" (mode: {used})"
    if payload.get("degraded_from"):
        header += f" [degraded from {payload['degraded_from']}]"
    lines = [header]
    for hit in hits:
        desc = f" — {hit['description']}" if hit.get("description") else ""
        why = f"  ({hit['why']})" if hit.get("why") else ""
        lines.append(f"  {hit['path']}  {hit['title']}{desc}{why}")
        if hit.get("snippet"):
            lines.append(f"      ↳ {hit['snippet']}")
    if not hits:
        lines.append("  (no hits — try `overview` for the whole brain)")
    return "\n".join(lines)


def _present_resolution_miss(payload: dict) -> str | None:
    """The disambiguation / not-found shapes shared by read and neighbors."""
    if "disambiguation" in payload:
        lines = ["several docs match — name one exactly:"]
        lines += [f"  {d['path']}  {d['title']}" for d in payload["disambiguation"]]
        return "\n".join(lines)
    if "error" in payload:
        lines = [payload["error"]]
        if payload.get("suggestions"):
            lines.append("did you mean:")
            lines += [f"  {s}" for s in payload["suggestions"]]
        return "\n".join(lines)
    return None


def present_read(payload: dict) -> str:
    miss = _present_resolution_miss(payload)
    if miss is not None:
        return miss
    frontmatter = payload.get("frontmatter") or {}
    title = frontmatter.get("title")
    heading = f"{payload['path']} — {title}" if title else payload["path"]
    lines = [heading, ""]
    lines.append((payload.get("content") or "").rstrip("\n"))
    neighbors = payload.get("neighbors", {})
    incoming = ", ".join(n["path"] for n in neighbors.get("in", []))
    outgoing = ", ".join(n["path"] for n in neighbors.get("out", []))
    if incoming or outgoing:
        lines.append("")
        lines.append(f"neighbors: in [{incoming}] out [{outgoing}]")
    if payload.get("truncated"):
        lines.append("")
        lines.append("(content trimmed — pass --json or read specific sections for the rest)")
    return "\n".join(lines)


def present_show(payload: dict) -> str:
    """The brainpick show result (spec/95): what the live server broadcast."""
    if "error" in payload:
        return payload["error"]
    shown = payload.get("shown", 0)
    seq = payload.get("seq")
    lines = [f"presented (seq {seq}): {shown} node(s) spotlighted in every open UI"]
    dropped = payload.get("dropped") or []
    if dropped:
        lines.append(f"dropped (unresolved): {', '.join(dropped)}")
    return "\n".join(lines)


def present_neighbors(payload: dict) -> str:
    miss = _present_resolution_miss(payload)
    if miss is not None:
        return miss
    header = f"neighbors of {payload['center']}"
    if payload.get("degraded_from"):
        header += f" [degraded from {payload['degraded_from']}]"
    lines = [header]
    others = [n for n in payload.get("nodes", []) if n["path"] != payload["center"]]
    for node in others:
        desc = f" — {node['description']}" if node.get("description") else ""
        lines.append(f"  d{node['distance']}  {node['path']}  {node['title']}{desc}")
    if not others:
        lines.append("  (no neighbors — an orphan, or nothing links here yet)")
    edges = payload.get("edges", [])
    if edges:
        lines.append("edges:")
        lines += [f"  {e['source']} → {e['target']} ({e['kind']})" for e in edges]
    return "\n".join(lines)
